pkg/guestagent/docker/events.go

```diff
@@ -0,0 +1,228 @@
+package docker
+
+import (
+	"context"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/events"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+	"github.com/lima-vm/lima/pkg/guestagent/api"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	startEvent = "start"
+	stopEvent  = "stop"
+	// die event is a confirmation of kill event.
+	dieEvent = "die"
+)
+
+type EventMonitor struct {
+	dockerClient *client.Client
+	// We maintain a record of all active containers because neither the stop nor
+	// die events provide the port mapping. As API consumers, it's our responsibility
+	// to track this information. The map uses the container ID as the key and
+	// stores all published ports associated with that container as the value.
+	runningContainers    map[string][]*api.IPPort
+	runningContainersMux sync.Mutex
+}
+
+func NewEventMonitor() *EventMonitor {
+	return &EventMonitor{
+		runningContainers: make(map[string][]*api.IPPort),
+	}
+}
+
+func (e *EventMonitor) MonitorPorts(ctx context.Context, ch chan *api.Event) {
+	e.tryGetClient(ctx)
+	defer e.dockerClient.Close()
+
+	if err := e.initializeRunningContainers(ctx, ch); err != nil {
+		logrus.Errorf("failed to initialize existing docker container published ports: %v", err)
+	}
+
+	msgCh, errCh := e.dockerClient.Events(ctx, events.ListOptions{
+		Filters: filters.NewArgs(
+			filters.Arg("type", "container"),
+			filters.Arg("event", startEvent),
+			filters.Arg("event", stopEvent),
+			filters.Arg("event", dieEvent)),
+	})
+
+	for {
+		select {
+		case <-ctx.Done():
+			logrus.Errorf("context cancellation: %v", ctx.Err())
+
+			return
+		case event := <-msgCh:
+			container, err := e.dockerClient.ContainerInspect(ctx, event.ID)
+			if err != nil {
+				logrus.Errorf("inspecting container [%v] failed: %v", event.ID, err)
+
+				continue
+			}
+
+			portMap := container.NetworkSettings.NetworkSettingsBase.Ports
+			logrus.Infof("received an event: {Status: %+v ContainerID: %+v Ports: %+v}",
```

Comment:
I think this might be a Debug log entry.

My criteria is: would I ever need to see this as a user, or do I only ever need to see this when I'm debugging why something isn't working? If the information isn't useful for a regular user, then it should be Debug and not Info.